src/main/java/cam72cam/immersiverailroading/registry/DefinitionManager.java

```diff
@@ -234,6 +242,41 @@ private static JsonObject getJsonData(String defID) throws IOException {
         return result;
     }
 
+    /**
+     * Get a stream for a collection that is used to load stocks in a singlethreaded or a multithreaded way.
+     *
+     * @param collection Collection of items.
+     * @param <E> Type of item.
+     * @return Singlethreaded or multithreaded stream.
+     */
+    private static <E> Stream<E> getStockLoadingStream(Collection<E> collection) {
+        if (!ConfigPerformance.multithreadedStockLoading) {
+            return collection.stream();
+        }
+
+        // Parallel streams use numCPUs-1 threads for stream workloads.
+        Runtime runtime = Runtime.getRuntime();
+        int processors = runtime.availableProcessors() - 1;
+        if (processors <= 1) {
+            return collection.stream();
+        }
+
+        // Manual garbage collection so we get an accurate quantity of free memory.
+        runtime.gc();
+
+        long maxMemory = runtime.maxMemory();
+        long totalMemory = runtime.totalMemory();
+        if (maxMemory == Long.MAX_VALUE) {
+            maxMemory = totalMemory;
+        }
+
+        long freeMemory = runtime.freeMemory();
+        freeMemory += maxMemory - totalMemory;
```

Comment:
I don't understand this part?  Would this double the amount of detected free memory?

Reply:
So for example when you have an initial heap size (Xms) of 1GB and a maximum heap size (Xmx) of 2GB, when this code gets the amount of free memory, it can be from a total anywhere between 1GB and 2GB. When more memory is needed the heap size gets increased until it hits 2GB.

So to get the true amount of free memory you need to know how big the heap currently is (totalMemory) and to what size it can grow (maxMemory). You then add the size difference to the indicated freeMemory, because when growing the heap, that extra memory becomes available to use.

Reply:
Ah, thanks for the explanation!